src/foam/dao/MongoDAO.java

```diff
@@ -0,0 +1,176 @@
+      MongoCredential credential = MongoCredential.createCredential(username, dbName, password.toCharArray());
+      mongoClient = new MongoClient(new ServerAddress(host, port), Arrays.asList(credential));
+    } else {
+      mongoClient = new MongoClient(host, port);
+    }
+
+    this.database = mongoClient.getDatabase(dbName);
+  }
+
+  public MongoDAO(String dbName, String username, String password) {
+    this("localhost", 27017, dbName, username, password);
+  }
+
+  public MongoDAO(String dbName) {
+    this(dbName, null, null);
+  }
+
+  private boolean isUserPassProvided(String username, String password) {
+    return ( ( username != null ) && ( ! username.isEmpty() ) ) &&
+            ( ( password != null ) && ( ! password.isEmpty() ) );
+  }
+
+  @Override
+  public Sink select_(X x, Sink sink, long skip, long limit, Comparator order, Predicate predicate) {
+    if ( sink == null ) {
+      sink = new ListSink();
+    }
+
+    Sink         decorated = decorateSink_(sink, skip, limit, order, predicate);
+    Subscription sub       = new Subscription();
+
+    String collectionName = (String) x.get("collectionName");
+    String collectionClass = (String) x.get("collectionClass");
+    List props = getFObjectProperties(collectionClass, x);
+
+    if ( collectionName == null || collectionName.isEmpty() ) {
+      throw new IllegalArgumentException("Invalid collection name in context. Please provide a string name.");
+    } else if ( collectionClass == null || collectionClass.isEmpty() ) {
+      throw new IllegalArgumentException("Invalid collection class name in context. Please provide a string class name.");
+    }
+
+    MongoCollection<Document> collection = database.getCollection(collectionName);
+    MongoCursor<Document> cursor = collection.find().iterator();
+
+    try {
+      while ( cursor.hasNext() ) {
+        if ( sub.getDetached() ) break;
+
+        FObject obj = createFObject(collectionClass, props, cursor.next());
+
+        if ( ( predicate == null ) || predicate.f(obj) ) {
+          decorated.put(obj, sub);
+        }
+      }
+    } finally {
+      cursor.close();
+    }
+
+    decorated.eof();
+
+    return sink;
+  }
+
+  private List getFObjectProperties(String clsName, X x) {    
+    try {
+      Class cls = Class.forName(clsName);
+      FObject clsInstance = (FObject) x.create(cls);
+      List props = clsInstance.getClassInfo().getAxiomsByClass(foam.core.AbstractFObjectPropertyInfo.class);
+      
+      // Recursively handle nested FObjectProperties
+      for ( int i = 0 ; i < props.size() ; i++ ) {
+        props.addAll(getFObjectProperties(((PropertyInfo) props.get(i)).getPropertyType(), x));
+      }
+
+      return props;
+    } catch (Exception ex) {
+      throw new RuntimeException(clsName + " was not found.");
+    }
+  }
+
+  private FObject createFObject(String clsName, List props, Document d) {
+    JsonWriterSettings writerSettings = new JsonWriterSettings(JsonMode.SHELL, true);
```

Comment:
I've updated the method to parse the BSON document directly instead of the previous technique. This leverages the internal mongodb-driver APIs, and doesn't require the JSON parsers at all. Note, I've reverted the updates to the StringParser and DateParser, but kept the update to the BooleanParser (as that may be useful elsewhere as well). 

I've implemented it such that it creates an object from context, and then parses the BSON Document fields, setting the value of individual properties. This will also handle nested objects well.